damien/lib/queries.py

```diff
@@ -24,40 +24,38 @@
 """
 
 import os
+import re
 
 from damien import db
-from damien.lib.util import resolve_sql_template_string
 from flask import current_app as app
 from sqlalchemy.sql import text
 
 
 def get_loch_basic_attributes(id_snippet, limit=20, exclude_uids=None):
     if os.environ.get('DAMIEN_ENV') == 'test':
         return []
-    query = """SELECT * FROM dblink('{dblink_nessie_rds}',$NESSIE$
+    params = {'id_snippet': f'{id_snippet}%', 'limit': limit}
+    uid_exclusion = ''
+    if exclude_uids:
+        params['uids'] = exclude_uids
+        uid_exclusion = 'AND NOT ldap_uid = ANY(:uids)'
+    query = f"""SELECT * FROM dblink('{app.config['DBLINK_NESSIE_RDS']}',$NESSIE$
                 SELECT ldap_uid, sid, first_name, last_name, email_address
                   FROM sis_data.basic_attributes
+                  WHERE (ldap_uid LIKE :id_snippet OR sid LIKE :id_snippet)
+                    {uid_exclusion}
+                  LIMIT :limit
```

Comment:
At least locally, this seems to fix the dblink performance problem. The trick is that the filter and limit have to go inside the $NESSIE$ dblink clause, not outside. Otherwise the entire contents of the sis_data.basic_attributes table get returned from Nessie RDS to bugenhagen, and are only filtered down on the bugenhagen side.